core/src/main/java/org/apache/calcite/sql/dialect/BigQuerySqlDialect.java

```diff
@@ -583,44 +582,32 @@ private SqlNode makCastCallForTimeWithPrecision(SqlNode operandToCast, int preci
 
   private void unparseRegexSubstr(SqlWriter writer, SqlCall call, int leftPrec, int rightPrec) {
     SqlCall extractCall;
-    switch (call.operandCount()) {
-    case 3:
-      extractCall = makeExtractSqlCall(call);
-      REGEXP_EXTRACT.unparse(writer, extractCall, leftPrec, rightPrec);
-      break;
-    case 4:
-    case 5:
-      extractCall = makeExtractSqlCall(call);
-      REGEXP_EXTRACT_ALL.unparse(writer, extractCall, leftPrec, rightPrec);
-      writeOffset(writer, call);
-      break;
-    default:
-      REGEXP_EXTRACT.unparse(writer, call, leftPrec, rightPrec);
-    }
-  }
-
-  private void writeOffset(SqlWriter writer, SqlCall call) {
-    int occurrenceNumber = Integer.parseInt(call.operand(3).toString()) - 1;
-    writer.literal("[OFFSET(" + occurrenceNumber + ")]");
+    extractCall = makeExtractSqlCall(call);
+    REGEXP_SUBSTR.unparse(writer, extractCall, leftPrec, rightPrec);
   }
 
   private SqlCall makeExtractSqlCall(SqlCall call) {
-    SqlCall substringCall = makeSubstringSqlCall(call);
-    call.setOperand(0, substringCall);
-    if (call.operandCount() == 5 && call.operand(4).toString().equals("'i'")) {
+    if (call.operandCount() == 5 || call.operand(1).toString().contains("\\")) {
       SqlCharStringLiteral regexNode = makeRegexNode(call);
       call.setOperand(1, regexNode);
     }
-    SqlNode[] extractNodeOperands = new SqlNode[]{call.operand(0), call.operand(1)};
-    return new SqlBasicCall(REGEXP_EXTRACT, extractNodeOperands, SqlParserPos.ZERO);
+    SqlNode[] extractNodeOperands;
+    if (call.operandCount() == 5) {
+      extractNodeOperands = new SqlNode[]{call.operand(0), call.operand(1),
+          call.operand(2), call.operand(3)};
+    } else {
+      extractNodeOperands = call.getOperandList().toArray(new SqlNode[0]);
+    }
+    return new SqlBasicCall(REGEXP_SUBSTR, extractNodeOperands, SqlParserPos.ZERO);
   }
 
   private SqlCharStringLiteral makeRegexNode(SqlCall call) {
-    String regexStr = call.operand(1).toString();
-    regexStr = regexStr.replace("\\", "\\\\");
-    String regexLiteral = "(?i)".concat(regexStr.substring(1, regexStr.length() - 1));
-    return SqlLiteral.createCharString(regexLiteral,
-        call.operand(1).getParserPosition());
+    String regexLiteral = call.operand(1).toString();
+    regexLiteral = regexLiteral.replace("\\", "\\\\");
+    if (call.operandCount() == 5 && call.operand(4).toString().equals("'i'")) {
+      regexLiteral = "(?i)".concat(regexLiteral.substring(1, regexLiteral.length() - 1));
+    }
+    return SqlLiteral.createCharString(regexLiteral, call.operand(1).getParserPosition());
```

Comment:
substring just once while returning

Reply:
Done with changes.Just check once

Reply:
you need to substring while returning. I don't see any code changes from previous version